File: services/30-ws-ruby/lib/jwt_service.rb
require 'jwt'

class JwtService
  ALGORITHM = 'HS256'.freeze
  TTL_SECONDS = 8 * 60 * 60 # 8 hours

  def self.encode(payload)
    exp = Time.now.to_i + TTL_SECONDS
    claims = payload.merge(exp: exp, iat: Time.now.to_i)
    JWT.encode(claims, secret, ALGORITHM)
  end

  def self.decode(token)
    decoded = JWT.decode(token, secret, true, algorithm: ALGORITHM)
    decoded.first.transform_keys(&:to_sym)
  rescue JWT::DecodeError, JWT::ExpiredSignature, JWT::VerificationError
    nil
  end

  def self.secret
    ENV.fetch('SECRET_KEY_BASE') do
      raise 'SECRET_KEY_BASE environment variable is not set'
    end
  end
  private_class_method :secret
end

File: services/28-php-grpc/src/Auth/JwtService.php
<?php

declare(strict_types=1);

namespace App\Auth;

use Firebase\JWT\JWT;
use Firebase\JWT\Key;

/**
 * JwtService — issues and validates JWT tokens for gRPC PHP service auth.
 *
 * JWT secret: loaded from JWT_SECRET env var.
 * Expiry: 8 hours from issuance.
 * Algorithm: HS256.
 *
 * Requires: firebase/php-jwt (composer require firebase/php-jwt)
 */
class JwtService
{
    private const ALGORITHM = 'HS256';
    private const TTL_SECONDS = 8 * 60 * 60; // 8 hours

    public static function encode(string $userId, string $email, string $name): string
    {
        $now = time();
        $payload = [
            'sub'   => $userId,
            'email' => $email,
            'name'  => $name,
            'iat'   => $now,
            'exp'   => $now + self::TTL_SECONDS,
        ];

        return JWT::encode($payload, self::secret(), self::ALGORITHM);
    }

    /**
     * Decodes and verifies a JWT string.
     *
     * Returns the decoded payload array on success.
     * Returns null on invalid signature, expiry, or malformed token.
     *
     * @return array<string, mixed>|null
     */
    public static function decode(string $token): ?array
    {
        try {
            $decoded = JWT::decode($token, new Key(self::secret(), self::ALGORITHM));
            return (array) $decoded;
        } catch (\Exception) {
            return null;
        }
    }

    private static function secret(): string
    {
        $secret = getenv('JWT_SECRET');
        if ($secret === false || $secret === '') {
            throw new \RuntimeException('JWT_SECRET environment variable is not set');
        }
        return $secret;
    }
}
